### PR TITLE
Added simulator bridge that publishes sub pos, orientation, and depth

### PR DESCRIPTION
**Pull request by dcrossman**

### CODE REVIEW

File: src/sensors/simulator_bridge.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// the simulator
+// Currently it publishes the position, orientation, and depth
+// of the sub
+// TODO: Eventually this should be generalized to publish a list
+// (defined in params) of objects as well as calulate depth
+void modelStatesCallback(const gazebo_msgs::ModelStates& msg)
+{
+    geometry_msgs::Vector3 position_msg;
+    geometry_msgs::Quaternion orientation_msg;
+    geometry_msgs::Vector3 lin_accel_msg;
```
Comment (skallaher) [inline src/sensors/simulator_bridge.cpp]: You commented out the publisher for the `lin_accel` shouldn't this also be commented out?

**Review by skallaher — APPROVED** [review]

**Comment by skallaher** [comment]
It looks like the Travis build failed because cmake couldn't find gazebo_msgs.

**Comment by dcrossman** [comment]
We'll need to add ros-indigo-gazebo7-msgs to the apt repo.

**Comment by dcrossman** [comment]
One more commit to review! I added a publisher for simulator object positions (useful for localization, ai, unit tests). Technically this is redundant since this information is found in the gazebo_msgs/ModelStates but it should simplify things a lot.

File: src/sensors/simulator_bridge.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    orientation_pub = nh.advertise<geometry_msgs::Quaternion>("orientation", 1);
+    depth_pub = nh.advertise<std_msgs::Float32>("depth", 1);
+    obstacle_pos_pub = nh.advertise<robosub::ObstaclePosArray>("obstacle_positions", 1);
+    // Info not being published right now
+    //lin_accel_pub = nh.advertise<geometry_msgs::Vector3>("lin_accel", 1);
+
+	ros::Subscriber orient_sub = nh.subscribe("gazebo/model_states", 1, modelStatesCallback);
+
+    int rate;
+    if(!nh.getParam("control/rate", rate))
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp:113]: Add brackers

File: src/sensors/simulator_bridge.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    ros::NodeHandle nh;
+
+    position_pub = nh.advertise<geometry_msgs::Vector3>("position", 1);
+    orientation_pub = nh.advertise<geometry_msgs::Quaternion>("orientation", 1);
+    depth_pub = nh.advertise<std_msgs::Float32>("depth", 1);
+    obstacle_pos_pub = nh.advertise<robosub::ObstaclePosArray>("obstacle_positions", 1);
+    // Info not being published right now
+    //lin_accel_pub = nh.advertise<geometry_msgs::Vector3>("lin_accel", 1);
+
+	ros::Subscriber orient_sub = nh.subscribe("gazebo/model_states", 1, modelStatesCallback);
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: unindent

File: src/sensors/simulator_bridge.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    ros::Rate r(rate);
+
+    // Put object names in vector
+    if(!nh.getParam("/obstacles", object_names))
+    {
+        ROS_WARN_STREAM("failed to load obstacle names");
+    }
+
+    // I use spinOnce and sleeps here because the simulator publishes
+    // at a very high rate and we don't need every message
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Period

File: src/sensors/simulator_bridge.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    // You can throttle subscriber input in other ways but
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: comma

File: src/sensors/simulator_bridge.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    // doesn't require extra packages or anything
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Period

File: src/sensors/simulator_bridge.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+#include "geometry_msgs/Vector3.h"
+#include "gazebo_msgs/ModelState.h"
+#include "gazebo_msgs/ModelStates.h"
+#include "robosub/ObstaclePosArray.h"
+
+ros::Publisher position_pub;
+ros::Publisher orientation_pub;
+ros::Publisher depth_pub;
+ros::Publisher obstacle_pos_pub;
+//ros::Publisher lin_accel_pub;
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Remove commented out lines of code  if you plan on adding them in future pull requests.

File: src/sensors/simulator_bridge.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+// the simulator
+// Currently it publishes the position, orientation, and depth
+// of the sub
+// TODO: Eventually this should be generalized to publish a list
+// (defined in params) of objects as well as calulate depth
+void modelStatesCallback(const gazebo_msgs::ModelStates& msg)
+{
+    geometry_msgs::Vector3 position_msg;
+    geometry_msgs::Quaternion orientation_msg;
+    //geometry_msgs::Vector3 lin_accel_msg;
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Remove

File: src/sensors/simulator_bridge.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    robosub::ObstaclePosArray object_array;
+    for(int i=0; i<object_names.size(); i++)
+    {
+        // Find object_names[i].
+        auto it = std::find(msg.name.begin(), msg.name.end(), object_names[i]);
+        // Note: std::find returns an iterator equal to array.end() if input
+        // not found in array.
+        if(it != msg.name.end())
+        {
+            // Calculate index
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Calculate index in array from the iterator element.

File: src/sensors/simulator_bridge.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            auto idx = std::distance(msg.name.begin(), it);
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Don't make this auto for clarity.

File: src/sensors/simulator_bridge.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            pos.z = msg.pose[idx].position.z;
+            pos.name = msg.name[idx];
+
+            object_array.data.push_back(pos);
+        }
+    }
+
+    obstacle_pos_pub.publish(object_array);
+
+    //lin_accel_pub.publish(lin_accel_msg);
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Remove commented out line and newline above.

File: src/sensors/simulator_bridge.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+int main(int argc, char **argv)
+{
+	ros::init(argc, argv, "simulator_bridge");
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Fix indentation.

File: src/sensors/simulator_bridge.cpp [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+    ros::NodeHandle nh;
+
+    position_pub = nh.advertise<geometry_msgs::Vector3>("position", 1);
+    orientation_pub = nh.advertise<geometry_msgs::Quaternion>("orientation", 1);
+    depth_pub = nh.advertise<std_msgs::Float32>("depth", 1);
+    obstacle_pos_pub = nh.advertise<robosub::ObstaclePosArray>("obstacle_positions", 1);
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: change topic to obstacles/positions

File: src/sensors/simulator_bridge.cpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    // Info not being published right now
+    //lin_accel_pub = nh.advertise<geometry_msgs::Vector3>("lin_accel", 1);
```
Comment (ryan-summers) [inline src/sensors/simulator_bridge.cpp]: Remove commented out line of code.

**Review by ryan-summers — CHANGES_REQUESTED** [review]
Remove commented lines of code and all unneeded newlines.

**Review by ryan-summers — APPROVED** [review]
Everything is looking good from here pending addition of the gazebo packages to the robosub packages.

**Pull request merged**